### PR TITLE
Enable upstream new conformance tests

### PR DESCRIPTION
**Pull request by slinkydeveloper**
Signed-off-by: Francesco Guardiani <francescoguard@gmail.com>

## Proposed Changes

- :gift: Add upstream new broker conformance tests to our test suite
- :bug: Fixed 405 Method not allowed conformance issue

### CODE REVIEW

**Comment by slinkydeveloper** [comment]
I would love to merge https://github.com/knative-sandbox/eventing-kafka-broker/pull/911 before this one

**Comment by slinkydeveloper** [comment]
The error seems in the retries (maybe some issue with the retry asserts?):

```
    --- FAIL: TestBrokerConformance/Knative_Broker_Specification_-_Control_Plane/Delivery_Spec_-_dlq-test-2/Assert/[Stable/SHOULD]Conformance_When_`TriggerSpec.Delivery`_is_configured,_then_`TriggerSpec.Delivery`_SHOULD_be_used._(Retry) (0.00s)
```

File: data-plane/dispatcher/src/main/java/dev/knative/eventing/kafka/broker/dispatcher/http/HttpConsumerVerticleFactory.java [hunk 1]
```diff
@@ -238,7 +238,7 @@ private static CircuitBreakerOptions createCircuitBreakerOptions(final DataPlane
         )
         // TODO max failures should be configurable or, at least, set by the control plane
         .setMaxFailures(egressConfig.getRetry() * 2)
-        .setMaxRetries(egressConfig.getRetry())
+        .setMaxRetries(egressConfig.getRetry() + 1)
```
Comment (matzew) [inline data-plane/dispatcher/src/main/java/dev/knative/eventing/kafka/broker/dispatcher/http/HttpConsumerVerticleFactory.java:241]: I am fine in doing the `+ 1` here. But should it work without? Perhaps reasonable to track as GH issue?

File: data-plane/receiver/src/main/java/dev/knative/eventing/kafka/broker/receiver/RequestMapper.java [hunk 2]
```diff
@@ -119,6 +121,16 @@ public void handle(final HttpServerRequest request) {
       return;
     }
 
+    if (request.method() != HttpMethod.POST) {
+      request.response().setStatusCode(METHOD_NOT_ALLOWED.code()).end();
+
+      logger.warn("method not allowed {}",
+        keyValue("method", request.method())
+      );
```
Comment (matzew) [inline data-plane/receiver/src/main/java/dev/knative/eventing/kafka/broker/receiver/RequestMapper.java:129]: spec says:

> Brokers MUST reject all HTTP produce requests with a method other than POST responding with HTTP status code `405 Method Not Supported`.

Perhaps we use that as some better `WARN` text?

**Comment by matzew** [comment]
looking here: https://prow.knative.dev/view/gs/knative-prow/pr-logs/pull/knative-sandbox_eventing-kafka-broker/1016/pull-knative-sandbox-eventing-kafka-broker-integration-tests/1407257040924971008

> FAIL: TestBrokerConformance/Knative_Broker_Specification_-_Control_Plane/Delivery_Spec_-_dlq-test-1/Assert/[Stable/SHOULD]Conformance_When_`BrokerSpec.Delivery`_is_configured,_but_not_the_specific_`TriggerSpec.Delivery`,_then_the_`BrokerSpec.Delivery`_SHOULD_be_used._(Retry) (0.00s)

**Comment by slinkydeveloper** [comment]
Interesting, now it fails on prow as well, let me try to revert the previous commit...

File: data-plane/receiver/src/main/java/dev/knative/eventing/kafka/broker/receiver/RequestMapper.java (hunk 2 above)
Comment (slinkydeveloper) [reply data-plane/receiver/src/main/java/dev/knative/eventing/kafka/broker/receiver/RequestMapper.java]: How about now?

File: data-plane/dispatcher/src/main/java/dev/knative/eventing/kafka/broker/dispatcher/http/HttpConsumerVerticleFactory.java (hunk 1 above)
Comment (slinkydeveloper) [reply data-plane/dispatcher/src/main/java/dev/knative/eventing/kafka/broker/dispatcher/http/HttpConsumerVerticleFactory.java]: removed the +1

File: data-plane/receiver/src/main/java/dev/knative/eventing/kafka/broker/receiver/RequestMapper.java [hunk 3]
```diff
@@ -124,7 +124,7 @@ public void handle(final HttpServerRequest request) {
     if (request.method() != HttpMethod.POST) {
       request.response().setStatusCode(METHOD_NOT_ALLOWED.code()).end();
 
-      logger.warn("method not allowed {}",
+      logger.warn("Only POST method is allowed. Method not allowed: {}",
```
Comment (matzew) [inline data-plane/receiver/src/main/java/dev/knative/eventing/kafka/broker/receiver/RequestMapper.java:127]: :+1:

**Review by matzew — APPROVED** [review]
/lgtm
/approve

**Comment by slinkydeveloper** [comment]
/hold

Let's wait until test pass consistently here

**Comment by matzew** [comment]
> 
    --- FAIL: TestBrokerConformance/Knative_Broker_Specification_-_Control_Plane/Delivery_Spec_-_dlq-test-2/Assert/[Stable/SHOULD]Conformance_When_`TriggerSpec.Delivery`_is_configured,_then_`TriggerSpec.Delivery`_SHOULD_be_used._(Retry) (0.00s)


on the `kind` e2e, I still see the above

**Comment by slinkydeveloper** [comment]
Yeah but it pass on prow...

**Comment by pierDipi** [comment]
Are they failing only with eventing v0.21? Perhaps that feature wasn't available at that time? 🤷‍♂️

**Comment by slinkydeveloper** [comment]
> Are they failing only with eventing v0.21? Perhaps that feature wasn't available at that time? man_shrugging

Uhm I don't know, retry should have been there, but maybe the overwrite trigger -> broker might be the missing bit in 0.21...

**Comment by matzew** [comment]
should we run aganst nightly?

**Comment by slinkydeveloper** [comment]
We run against latest as well (see 0.22). But simply, this test seems like it's failing with any < 0.22 releases

**Comment by matzew** [comment]
I meant more if we should only run against nightly and last releaes (or last two releases)?

**Comment by slinkydeveloper** [comment]
I think the policy of knative is to support the latest 3 major releases, that's the reason we test all these different versions.

**Comment by matzew** [comment]
> I think the policy of knative is to support the latest 3 major releases, that's the reason we test all these different versions.

next week we cut the 0.24 :-)   so:
* 0.22
* 0.23
* 0.24
* main
done

**Comment by pierDipi** [comment]
Let's remove eventing v0.21 since these tests are useful while we're refactoring code.

**Comment by slinkydeveloper** [comment]
Will do in this PR now

**Comment by slinkydeveloper** [comment]
Rebased and removed 0.21. I've also removed the new `RequestMapper` for the http method check, just because it was going to create even more conflicts with #1041 that is going to properly reorganize that code anyway.

**Comment by pierDipi** [comment]
https://github.com/knative-sandbox/eventing-kafka-broker/pull/1016/checks?check_run_id=2917490884 `Unit Tests (1.16.x, ubuntu-latest)`
```
 git remote add upstream https://github.com/knative-sandbox/eventing-kafka-broker.git
  git fetch upstream main
  git pull --no-rebase --no-commit upstream main
  shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
  env:
    GOROOT: /opt/hostedtoolcache/go/1.16.5/x64
From https://github.com/knative-sandbox/eventing-kafka-broker
 * branch              main       -> FETCH_HEAD
 * [new branch]        main       -> upstream/main
From https://github.com/knative-sandbox/eventing-kafka-broker
 * branch              main       -> FETCH_HEAD
Committer identity unknown

*** Please tell me who you are.

Run

  git config --global user.email "you@example.com"
```

**Comment by slinkydeveloper** [comment]
@pierDipi not this pr error at all :) I suggest we merge this one and then wait for the next actions PR fix...

File: .github/workflows/kind-e2e.yaml [hunk 4]
```diff
@@ -29,7 +29,6 @@ jobs:
           - "./third_party/eventing-latest/"
           - "./third_party/eventing-v0.23.0/"
           - "./third_party/eventing-v0.22.1/"
-          - "./third_party/eventing-v0.21.4/"
```
Comment (pierDipi) [inline .github/workflows/kind-e2e.yaml:32]: We need to remove this directory too

**Pull request merged**